[ARMv7] Set temporaryCallFrame in WebAssembly.asm

aoikonomopoulos opened the pull request:
#### fe797d6f6ba6dd16e331f7b0de522cf2a47a38b7
<pre>
[ARMv7] Set temporaryCallFrame in WebAssembly.asm
<a href="https://bugs.webkit.org/show_bug.cgi?id=294597">https://bugs.webkit.org/show_bug.cgi?id=294597</a>

Reviewed by Justin Michaud.

ARMv7 can&apos;t rely on BUILTIN_FRAME_ADDRESS (it&apos;s broken on GCC).
Properly initialize temporaryCallFrame in WebAssembly.asm.

This should really be `or (not USE_BUILTIN_FRAME_ADDRESS)`, but we need
to keep the number of configuration values down to avoid a combinatorial
explosion in the size of LLIntAssembly.h.

When we have a fix for GCC, we can just drop the reliance on
temporaryCallFrame, but in the meantime this avoids flaky tests in EWS.

* Source/JavaScriptCore/llint/WebAssembly.asm:

Canonical link: <a href="https://commits.webkit.org/298095@main">https://commits.webkit.org/298095@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42018132998801fc3eac7aafcd604d4b492d2f1e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114199 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33946 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24408 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120365 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64940 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34575 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42509 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86786 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41743 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117147 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27528 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102562 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67175 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26711 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20688 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64060 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/106610 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96901 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20805 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123584 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/112762 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41219 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30729 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95618 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41597 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98764 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95401 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24317 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40539 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18361 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41099 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46608 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/136959 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40708 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36661 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44012 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42461 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/llint/LLIntOfflineAsmConfig.h:
```diff
@@ -169,3 +169,9 @@
+#define OFFLINE_ASM_USE_BUILTIN_FRAME_ADDRESS 1
+#else
+#define OFFLINE_ASM_USE_BUILTIN_FRAME_ADDRESS 0
+#endif
```
This addition makes LLIntAssembly.h bloat significantly (since LLInt generates each combination).

Constellation commented on Source/JavaScriptCore/llint/WebAssembly.asm:
```diff
@@ -1159,7 +1158,7 @@ end
     copyCalleeSavesToVMEntryFrameCalleeSavesBuffer(a0, a1)
 
-if ASSERT_ENABLED
+if (not USE_BUILTIN_FRAME_ADDRESS) or ASSERT_ENABLED
```
I think that we should just make ARMv7 work with `BUILTIN_FRAME_ADDRESS`. It is the last JIT architecture which is not supporting this `BUILTIN_FRAME_ADDRES`.

Constellation replied:
In IPInt side, I think we should just put this frame without ASSERT_ENABLED. We would like to support arbitrary arch in IPInt, and we may support things without USE_BUILTIN_FRAME_ADDRESS. But anyway, I think definitely all JIT arch should support USE_BUILTIN_FRAME_ADDRESS

Constellation replied:
But for this particular case, I think we should just enable USE_BUILTIN_FRAME_ADDRESS for ARMv7. That's the simplest solution.

aoikonomopoulos replied:
Agreed. However it looks like we get broken code when doing a Release build with GCC on ARMv7. I can see it doing

```
sub     sp, #180        @ 0xb4
add     r7, sp, #24
```

and then dereferencing that to get the caller frame address. `clang` correctly sets up `r7` before allocating stack space. Looking for a workaround.

aoikonomopoulos replied:
I don't see any way to work around this, I think this needs an upstream GCC fix. How about we gate this on `ARMv7` in WebAssembly.asm for now?

aoikonomopoulos replied:
I.e. to avoid the blow up in `LLIntAssembly.h`.

aoikonomopoulos committed:
[ARMv7] Set temporaryCallFrame in WebAssembly.asm
https://bugs.webkit.org/show_bug.cgi?id=294597

Reviewed by Justin Michaud.

ARMv7 can't rely on BUILTIN_FRAME_ADDRESS (it's broken on GCC).
Properly initialize temporaryCallFrame in WebAssembly.asm.

This should really be `or (not USE_BUILTIN_FRAME_ADDRESS)`, but we need
to keep the number of configuration values down to avoid a combinatorial
explosion in the size of LLIntAssembly.h.

When we have a fix for GCC, we can just drop the reliance on
temporaryCallFrame, but in the meantime this avoids flaky tests in EWS.

* Source/JavaScriptCore/llint/WebAssembly.asm:

Canonical link: https://commits.webkit.org/298095@main
